homeassistant/components/websocket_api/messages.py

```diff
@@ -74,6 +74,104 @@ def _cached_event_message(event: Event) -> str:
     return message_to_json(event_message(IDEN_TEMPLATE, event))
 
 
+def cached_state_diff_message(iden: int, event: Event) -> str:
+    """Return an event message.
+
+    Serialize to json once per message.
+
+    Since we can have many clients connected that are
+    all getting many of the same events (mostly state changed)
+    we can avoid serializing the same data for each connection.
+    """
+    return _cached_state_diff_message(event).replace(IDEN_JSON_TEMPLATE, str(iden), 1)
+
+
+@lru_cache(maxsize=128)
+def _cached_state_diff_message(event: Event) -> str:
+    """Cache and serialize the event to json.
+
+    The IDEN_TEMPLATE is used which will be replaced
+    with the actual iden in cached_event_message
+    """
+    return message_to_json(event_message(IDEN_TEMPLATE, _state_diff_event(event)))
+
+
+def _state_diff_event(event: Event) -> dict:
+    """Convert a state_changed event to the minimal version.
+
+    State update example
+
+    {
+        add: {entity_id: state,…}
+        changed: {entity_id: diff,…}
+        removed: [entity_id,…]
+    }
+
+    Init state should do as_dict, copy, remove entity id since it will be in the key
+
+    Fetch function is empty
+    """
+    # The entity_id is also duplicated in the message twice but its actually used
+    if (event_new_state := event.data["new_state"]) is None:
+        return {"remove": [event.data["entity_id"]]}
+    assert isinstance(event_new_state, State)
+    if (event_old_state := event.data["old_state"]) is None:
+        return {
+            "add": {
+                event_new_state.entity_id: compressed_state_dict_add(event_new_state)
+            }
+        }
+    assert isinstance(event_old_state, State)
+    return _state_diff(event_old_state, event_new_state)
+
+
+def _state_diff(
+    old_state: State, new_state: State
+) -> dict[str, dict[str, dict[str, dict[str, str | list[str]]]]]:
+    """Create a diff dict that can be used to overlay changes."""
+    old_state_dict = old_state.as_compressed_dict()
+    new_state_dict = new_state.as_compressed_dict()
+    diff: dict = {}
+    for item, value in new_state_dict.items():
```

Comment:
I'm not sure I like this design. We have known like for like data.  This is treating it as unknown data.  I think this can be a lot more efficient if we make it a bit less generic

Reply:
Yeah, that's what I did on the JS side.